File: xbmcontrol-evo/xbmcontrol-evo/XBMC.Library/XBMC.Playlist.cs
// ------------------------------------------------------------------------
//    XBMControl - A compact remote controller for XBMC (.NET 3.5)
//    Copyright (C) 2008  Bram van Oploo (bramvano@gmail.com)
//                        Mike Thiels (Mike.Thiels@gmail.com)
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
// ------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace XBMC
{
    public class XBMC_Playlist
    {
        XBMC_Communicator parent;
        private string[] aCurrentPlaylist = null;

        public XBMC_Playlist(XBMC_Communicator p)
        {
            parent = p;
        }

        public string[] Get(string type, bool parse)
        {
            string[] aPlaylistTemp = parent.Request("GetPlaylistContents(" +type+ ")");

            if (parse)
            {
                if (aPlaylistTemp != null)
                {
                    aCurrentPlaylist = new string[aPlaylistTemp.Length];
                    for (int x = 0; x < aPlaylistTemp.Length; x++)
                    {
                        int i = aPlaylistTemp[x].LastIndexOf(".");
                        if (i > 1)
                        {
                            string extension = aPlaylistTemp[x].Substring(i, aPlaylistTemp[x].Length - i);
                                aPlaylistTemp[x] = aPlaylistTemp[x].Replace("\\", "/");
                            string[] aPlaylistEntry = aPlaylistTemp[x].Split('/');
                            string playlistEntry = aPlaylistEntry[aPlaylistEntry.Length - 1].Replace(extension, "");
                            aCurrentPlaylist[x] = playlistEntry;
                        }
                        else
                            aCurrentPlaylist[x] = "";
					}
					
					return aCurrentPlaylist;
                }
				else
					return null;
            }
            else
                return aPlaylistTemp;
        }

        public bool PlaySong(int position)
        {
            string[] response = parent.Request("SetPlaylistSong(" + position.ToString() + ")");
			return parent.CreateBoolRespose(response);
        }

        public bool Remove(int position)
        {
            string[] response = parent.Request("RemoveFromPlaylist(" + position.ToString() + ")");
			return parent.CreateBoolRespose(response);
        }

        public string GetCurrentPlaylistType()
        {
            string[] response = parent.Request("GetCurrentPlaylist()");
            return (response == null) ? null : response[0];
        }

        public bool Clear()
        {
            string[] response = parent.Request("ClearPlayList()");
			return parent.CreateBoolRespose(response);
        }

        public bool AddDirectoryContent(string folderPath, string mask, bool recursive)
        {
            string p = "";
            string m = "";
            string r = "";
			
            if (mask != null)
            {
				p = ";0";
                m = ";[" + mask + "]";
                r = (recursive) ? ";1" : ";0";
            }

            string[] response = parent.Request("AddToPlayList(" + folderPath + p + m + r + ")");
			return parent.CreateBoolRespose(response);
        }

        public bool AddDirectoryContent(string folderPath, string mask)
        {
            return this.AddDirectoryContent(folderPath, mask, false);
        }

        public bool AddFilesToPlaylist(string filePath)
        {
            return this.AddDirectoryContent(filePath, null);
        }
		
		public int GetLength()
		{
			string[] length = parent.Request("GetPlaylistLength(0)");
			return (length == null)? 0 : Convert.ToInt32(length[0]);
		}

        public bool SetSong(int position)
        {
            string[] response = parent.Request("SetPlaylistSong(" + position.ToString() + ")");
			return parent.CreateBoolRespose(response);
        }

        public bool Set(string type)
        {
            string[] response = parent.Request("SetCurrentPlaylist(" + type + ")");
			return parent.CreateBoolRespose(response);
        }
    }
}

File: trunk/Java/XbmcLibrary/src/Xbmc/XbmcPlaylist.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package Xbmc;

import java.util.List;

/**
 *
 * @author Bram
 */
public class XbmcPlaylist {

    private XbmcConnection o_Xbmc               = null;
    private List<String> l_currentContents      = null;
    private List<String> l_currentContentNames  = null;
    private int i_contentsRequestCount          = 0;
    private int i_contentnamesRequestCount      = 0;

    public XbmcPlaylist (XbmcConnection o_connection) {
        o_Xbmc = o_connection;
    }

    public List<String> getContents (String s_type) {
        l_currentContents       = o_Xbmc.Request.send("GetPlaylistContents(" + s_type + ")");
        l_currentContentNames   = null;
        i_contentsRequestCount++;
        
        for (int x=0; x<l_currentContents.size(); x++) {
            int i = l_currentContents.get(x).lastIndexOf(".");
            if (i>1) {
                String s_extension          = l_currentContents.get(x).substring(i, (l_currentContents.get(x).length()-i));
                l_currentContents.set(x, l_currentContents.get(x).replace("\\", "/"));
                String[] sa_paylistEntry    = l_currentContents.get(x).split("/");
                String s_playlistEntry      = sa_paylistEntry[sa_paylistEntry.length-1].replace(s_extension, "");
                l_currentContentNames.add(s_playlistEntry);
            } else {
                l_currentContentNames.add("");
            }
        }

        return l_currentContents;
    }

    public List<String> getContentNames (String s_type) {
        i_contentnamesRequestCount++;
        if (i_contentsRequestCount != i_contentnamesRequestCount) getContents(s_type);

        return l_currentContentNames;
    }

    public Boolean playSong (int i_position) {
        List<String> l_response = o_Xbmc.Request.send("SetPlaylistSong(" + Integer.toString(i_position) + ")");
        return o_Xbmc.Request.getBooleanResponse(l_response.get(0));
    }

    public Boolean remove (int i_position) {
        List<String> l_response = o_Xbmc.Request.send("RemoveFromPlaylist(" + Integer.toString(i_position) + ")");
        return o_Xbmc.Request.getBooleanResponse(l_response.get(0));
    }

    public String getCurrentPlaylistType () {
        List<String> l_response = o_Xbmc.Request.send("GetCurrentPlaylist()");
        return (l_response == null)? null : l_response.get(0);
    }

    public Boolean clear () {
        List<String> l_response = o_Xbmc.Request.send("ClearPlayList()");
        return o_Xbmc.Request.getBooleanResponse(l_response.get(0));
    }

    public Boolean addDirectoryContent(String s_directoryPath, String s_mask, Boolean b_recursive) {
        String s_p = "";
        String s_m = "";
        String s_r = "";

        if (s_mask != null) {
            s_p = ";0";
            s_m = ";[" + s_mask + "]";
            s_r = (b_recursive)? ";1" : ";0";
        }

        List<String> l_response = o_Xbmc.Request.send("AddToPlayList(" + s_directoryPath + s_p + s_m + s_r + ")");
        return o_Xbmc.Request.getBooleanResponse(l_response.get(0));
    }

    public Boolean addDirectoryContent(String s_directoryPath, String s_mask)   { return addDirectoryContent(s_directoryPath, s_mask, false); }
    public Boolean addFilesToPlaylist(String s_filePath)                        { return addDirectoryContent(s_filePath, null, false); }

    public int getLength (String s_type) {
        List<String> l_playlistLength = o_Xbmc.Request.send("GetPlaylistLength(" + s_type + ")");
        return (l_playlistLength == null)? 0 : Integer.getInteger(l_playlistLength.get(0));
    }

    public Boolean setSong (int i_position) {
        List<String> l_response = o_Xbmc.Request.send("SetPlaylistSong(" + Integer.toString(i_position) + ")");
        return o_Xbmc.Request.getBooleanResponse(l_response.get(0));
    }

    public Boolean setType (String s_type) {
        List<String> l_response = o_Xbmc.Request.send("SetCurrentPlaylist(" + s_type + ")");
        return o_Xbmc.Request.getBooleanResponse(l_response.get(0));
    }
}
